File: 백준/Gold/7490. 0 만들기/0 만들기.cc
#include <iostream>
#include <string>
using namespace std;

bool zero(string s) {
    int sum = 0;
    int cur = 0;
    char c = '+';

    for (int i = 0; i < s.size(); i++) {
        if (isdigit(s[i]))
            cur = cur * 10 + (s[i] - '0');
        else if (s[i] == '+' or s[i] == '-') {
            if (c == '+')
                sum += cur;
            else if (c == '-')
                sum -= cur;

            if (s[i] == '+')
                c = '+';
            else if (s[i] == '-')
                c = '-';
            cur = 0;
        }

    }
    if (c == '+')
        sum += cur;
    else if (c == '-')
        sum -= cur;
    
    return (sum == 0);
}

char cal[3] = {' ' ,'+', '-'};
void dfs(int n, int cur, string s) {

    s += cur + '0';

    if (n <= cur) {
        if (zero(s))
            cout << s << "\n";
        return;
    }

    for (int i = 0; i < 3; i++) 
        dfs(n, cur + 1, s + cal[i]);

    
}


int main(void) {
    int t;
    cin >> t;
    while (t-- > 0) {
        int n;
        cin >> n;
        string s;
        dfs(n, 1, s);
        cout << "\n";
    }

}

File: 백준/Gold/7490. 0 만들기/0 만들기.java
import java.util.*;
import java.io.*;

public class Main{

    static int n;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int t = Integer.parseInt(br.readLine());
        while(t-- > 0){
            n = Integer.parseInt(br.readLine());
            dfs(2, "1");
            System.out.println();
        }

    }
    public static void dfs(int cur, String s){
        if(cur > n){
            if(zero(s))
                System.out.println(s);
            return;
        }
        dfs(cur + 1, s + " " + cur);
        dfs(cur + 1, s + "+" + cur);
        dfs(cur + 1, s + "-" + cur);
    }
    static boolean zero(String s){
        int val = 0;
        int x = 0;
        char ch = '+';
        for(int i = 0; i < s.length(); i++){
            if(s.charAt(i) == ' ') continue;
            if(Character.isDigit(s.charAt(i))){
                x = x * 10 + (s.charAt(i) - '0');
            }
            else{
                if(ch == '+') val += x;
                else val -= x;

                x = 0;
                ch = s.charAt(i);
            }
        }
        if(ch == '+') val += x;
        else val -= x;

        return (val == 0);
    }
}